### DISCUSSION TITLE
Zip indexed rows from 2d arrays

### DISCUSSION THREAD

**Discussion by vikigenius** [rust-ndarray/ndarray#996] [Q&A] [answered]
Say, I have two Array2 objects:

```rust
type M = Array2<f32>;
let mut a = M::eye(2);
let mut b = 2.0 * a;
```

I want to iterate over the rows of a and b in lockstep, but also have the row index simultaneously:

for example:
`(ridx, (row_a, row_b)) in Zip(a.rows(), b.rows()).enumerate()`

```
I should get

(0, ([1.0, 0], [2.0, 0]))
(1, ([0, 1.0], [0, 2.0]))
```
How can I do this efficiently? Also can i do the same thing in parallel (over row indices) ?

**Comment by bluss** [comment, marked as answer]
Zip::indexed can do this

**Reply by vikigenius** [reply]
Thanks, this works

```rust
Zip::indexed(a.outer_iter()).and(b.outer_iter()).into_par_iter().for_each(|(idx, ra, rb)| {
    println!("idx = {}, ra = {}, rb = {}", idx, ra, rb)
}) ;
```

Or much more concisely:

```rust
par_azip!((index idx, ra in a.rows(), rb in b.rows()) {
    println!("idx = {}, ra = {}, rb = {}", idx, ra, rb)
});
```

**Reply by bluss** [reply]
You can shorten `.into_par_iter().for_each()` to just `.par_for_each(|idx, ra, rb| ..)` at least. I've gone back and forth between preferring the macros or the methods, so it's the user's choice.
